bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -148,13 +177,70 @@ void ReplicaImp::messageHandler(MessageBase *msg) {
       }
     }
   }
-  if (!isCollectingState() && validateMessage(trueTypeObj)) {
+
+  if (!isCollectingState()) {
+    // The message validation of few messages require time-consuming processes like
+    // digest calculation, signature verification etc. Such messages are identified
+    // by review. During the review we also check if asynchronous message validation
+    // is feasible or not for the given message. After identification, true returning
+    // shouldValidateAsync() member function is added to these message which indicates
+    // that Asynchronous validation is possible for these messages.
+    // In this function we check shouldValidateAsync() and decide if asynchronous
+    // validation will happen or synchronous validation will happen.
+    // prePrepareFinalizeAsyncEnabled flag will be removed in 1.7, when this entire
+    // async behaviour will be assumed to be default for the chosen messages.
+    if (getReplicaConfig().prePrepareFinalizeAsyncEnabled && trueTypeObj->shouldValidateAsync()) {
+      asyncValidateMessage<T>(trueTypeObj);
+      return;
+    } else {
+      if (validateMessage(trueTypeObj)) {
+        onMessage<T>(trueTypeObj);
+        return;
+      }
+    }
+  }
+  delete trueTypeObj;
+}
+
+/**
+ * validatedMessageHandler<T> This is a family of validated message callback.
+ * As validation is assumed to happen, this function calls the onMessage<T> directly.
+ *
+ * @param msg : This is the carrier message which holds some message which was originally
+ *              external and now validated.
+ * @return : returns nothing
+ */
+template <typename T>
+void ReplicaImp::validatedMessageHandler(CarrierMesssage *msg) {
+  // The Carrier message is allocated by this replica and after validation its given back
+  // to itself without going through the network stack (as an internal message).
+  // So the carrier message is quickly reinterpreted and used instead of fresh memory allocation.
+  // So this reinterpret_cast is intended.
+  ValidatedMessageCarrierInternalMsg<T> *validatedTrueTyeObj =
+      reinterpret_cast<ValidatedMessageCarrierInternalMsg<T> *>(msg);
+  T *trueTypeObj = validatedTrueTyeObj->returnMessageToOwner();
+  delete msg;
+
+  if (bftEngine::ControlStateManager::instance().getPruningProcessStatus()) {
```

Comment:
can this block ever happen?
We already passed same check in ReplicaImp::messageHandler?
Maybe you should assert here?

Reply:
Pruning happens when all nodes are ready (its not based on quorum). During pruning, it sets a replica wide flag as during pruning we should never update the blockchain. So before starting the onMessage, the handler checks pruning flag and allows only client req messages and ignores all other protocol messages. As validation would have happened in a separate thread, both pruning and state transfer need to be checked.